src/SpectatorView.Unity/Assets/SpectatorView/Scripts/MarkerDetection/QRCode/QRCodesManager.cs

```diff
@@ -1,302 +1,337 @@
+                            newMatrix.M43 = -newMatrix.M43;
+
+                            newMatrix.M31 = -newMatrix.M31;
+                            newMatrix.M32 = -newMatrix.M32;
+                            newMatrix.M34 = -newMatrix.M34;
+
+                            System.Numerics.Vector3 winrtScale;
+                            System.Numerics.Quaternion winrtRotation;
+                            System.Numerics.Vector3 winrtTranslation;
+                            System.Numerics.Matrix4x4.Decompose(newMatrix, out winrtScale, out winrtRotation, out winrtTranslation);
+
+                            var translation = new Vector3(winrtTranslation.X, winrtTranslation.Y, winrtTranslation.Z);
+                            var rotation = new Quaternion(winrtRotation.X, winrtRotation.Y, winrtRotation.Z, winrtRotation.W);
+                            location = Matrix4x4.TRS(translation, rotation, Vector3.one);
+
+                            return true;
+                        }
+                        else
+                        {
+                            Debug.LogWarning("QRCode location unknown or not yet available.");
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        Debug.LogWarning("Failed to obtain coordinate system for application");
+                        return false;
+                    }
+                }
+                catch(Exception e)
+                {
+                    Debug.LogWarning($"Note: TryGetLocationForQRCode needs to be called from main thread: {e}");
+                    return false;
+                }
+            }
+            else
+            {
+                Debug.LogWarning("Failed to obtain coordinate system for QRCode");
+                return false;
+            }
+        }
+#endif // WINDOWS_UWP
+
+        public Guid GetIdForQRCode(string qrCodeData)
+        {
+            lock (qrCodesList)
+            {
+                foreach (var ite in qrCodesList)
+                {
+                    if (ite.Value.Code == qrCodeData)
+                    {
+                        return ite.Key;
+                    }
+                }
+            }
+            return new Guid();
+        }
+
+        public IList<QRCodesTrackerPlugin.QRCode> GetList()
+        {
+            lock (qrCodesList)
+            {
+                return new List<QRCodesTrackerPlugin.QRCode>(qrCodesList.Values);
+            }
+        }
+
+        void Awake()
+        {
+            IsTrackerRunning = false;
+        }
+
+        protected async void Start()
+        {
+            if (AutoStartQRTracking)
+            {
+                await StartQRTrackingAsync();
+            }
+        }
+
+        public async Task<QRTrackerStartResult> StartQRTrackingAsync()
+        {
+            bool attemptToStart = false;
+            lock (startupLock)

```

Comment:
>lock (startupLock) [](start = 12, length = 18)

Is this lock necessary? I think it's only necessary if something calls either of these methods from a different thread. If they're called from the Unity main thread, then the lock isn't necessary. Even after the awaited method returns the execution will still be on the Unity main thread due to its synchronization context.

Reply:
@andreiborodin and I chatted offline and he provided some great perspectives on how to handle this locking better. The code should now be updated.